### PR TITLE
Refactor timer to avoid deadlocks

### PR DESCRIPTION
**Pull request by anosach-luxoft**
* Change `state_lock_` variable name to `termination_lock_` to avoid name duplication
* Rename `StopUnsafe()` method to more clear `StopLockFree()`
* Remove meanless `task_lock_` (task cannot be deleted during running by design)
* **Do not set `stop_flag_` to false in `threadMain()` (actually fix)**
* Extend UT coverage of timer

@AGaliuzov @AGritsevich @Kozoriz @OHerasym @LuxoftAKutsan please review

### CODE REVIEW

File: src/components/utils/test/timer_test.cc [hunk 1]
```diff
@@ -43,214 +43,276 @@ namespace components {
...
+  timer::Timer timer(kTimerName, mock_task);
+
+  timer.Start(kDefaultTimeout, kSingleShot);
+  EXPECT_TRUE(timer.is_running());
+  EXPECT_EQ(kDefaultTimeout, timer.timeout());
+
+  timer.Start(kDefaultTimeoutRestart, kSingleShot);
+  EXPECT_TRUE(timer.is_running());
+  EXPECT_EQ(kDefaultTimeoutRestart, timer.timeout());
 }
```
Comment (Kozoriz) [inline src/components/utils/test/timer_test.cc]: @anosach-luxoft , can you call Stop in the end of test? (just for safety)

File: src/components/utils/test/timer_test.cc [hunk 2]
```diff
@@ -43,214 +43,276 @@ namespace components {
...
+
+  timer.Start(kDefaultTimeout, !kSingleShot);
+  EXPECT_TRUE(timer.is_running());
+  EXPECT_EQ(kDefaultTimeout, timer.timeout());
 
-  EXPECT_FALSE(test_timer.is_running());
+  timer.Start(kDefaultTimeoutRestart, !kSingleShot);
+  EXPECT_TRUE(timer.is_running());
+  EXPECT_EQ(kDefaultTimeoutRestart, timer.timeout());
 }
```
Comment (Kozoriz) [inline src/components/utils/test/timer_test.cc]: @anosach-luxoft , can you call Stop in the end of every Test ? (for safety)

Comment (anosach-luxoft) [inline src/components/utils/test/timer_test.cc]: @Kozoriz we don't need to call `Stop()` here. Timer is local object and it will be destroyed immediately.

**Comment by anosach-luxoft** [comment]
@AGaliuzov @AGritsevich @Kozoriz @OHerasym @LuxoftAKutsan please review

File: src/components/utils/include/utils/timer.h [hunk 3]
```diff
@@ -141,42 +141,30 @@ class Timer {
...
      */
-    sync_primitives::Lock state_lock_;
+    sync_primitives::Lock termination_lock_;
     sync_primitives::ConditionalVariable termination_condition_;
 
     DISALLOW_COPY_AND_ASSIGN(TimerDelegate);
   };
 
-  void StopUnsafe();
+  void StopLockFree();
```
Comment (AGaliuzov) [inline src/components/utils/include/utils/timer.h]: I think the name `Stop` is quite enough here. What for we need to specifically add `LockFree`?

Comment (anosach-luxoft) [inline src/components/utils/include/utils/timer.h]: @AGaliuzov because we already have `Stop()` method that locks mutex as opposed to this one

Comment (AGaliuzov) [inline src/components/utils/include/utils/timer.h]: @anosach-luxoft so may be we also need re-name the opposite `Stop` into `StopWithLock`

**Comment by AGaliuzov** [comment]
Reviewed

File: src/components/utils/include/utils/timer.h (hunk 3 above)
Comment (anosach-luxoft) [inline src/components/utils/include/utils/timer.h]: @AGaliuzov no. How do you see such interface? `Start()` and `StopWithLock()`?? or `StartWithLock()` and `StopWithLock()`?

**Comment by Kozoriz** [comment]
Reviewed

**Comment by AGritsevich** [comment]
Reviewed

**Comment by anosach-luxoft** [comment]
Closed in favor of #503

**Pull request closed without merging**